invoke/runners.py

```diff
@@ -322,8 +322,22 @@ def get():
                     # Can't use six.b because that just assumes latin-1 :(
                     data = data.encode(self.encoding)
                 yield data
-        # Decode stream using our generator & requested encoding
-        for data in codecs.iterdecode(get(), self.encoding, errors='replace'):
+        if not hasattr(output, 'encoding') or self.encoding == output.encoding:
+            # No need in re-encoding if output is a BytesIO or CarbonCopy, or
+            # source and target encodings are the same
+            data_stream = get()
+        else:
+            # Re-encode from user requested encoding to the output encoding
+            data_stream = codecs.iterencode(
+                codecs.iterdecode(
+                    get(),
+                    self.encoding,
+                    errors='replace'
+                ),
+                output.encoding,
+                errors='replace'
+            )
+        for data in data_stream:
```

Comment:
OK. I've looked at the surrounding code. I think I see what you're trying to achieve here.

You're right, there's a bug, but your fix is wrong (see below - I'm taking a purist view here, I suspect your fix probably works, but it's very hard to analyze, so I can't be sure). According to the comment at the top of the ```io()``` function, ```reader``` should return bytes, and ```output``` should take bytes as the argument to its ```write()``` method. In Python 3, that's explicitly typed, and using the wrong types won't work. In Python 2, you can accidentally supply a string when you mean bytes and it will work, but give the wrong output. This is where Python 2's string/bytes model sucks - you have to track which objects are bytes and which are strings for yourself, the language doesn't do it for you.

The original code got bytes from ```get()```, put them through ```iterdecode``` to get strings, then wrote the strings with ```output.write```. That is broken, because you're writing strings to an output function that should get bytes.

But the comment about ```output``` is muddled:

> If ``hide`` is ``False``, write bytes to ``output``, a stream such as `sys.stdout`.

That doesn't make sense because ```sys.stdout``` is not a stream you can write bytes to! You can write bytes to the test CarbonCopy stream, so the tests don't spot this. But you *cannot* supply bytes to ```sys.stdout.write```. On Python 2 you get mojibake, on Python 3 you get a type error.

I *think* your code juggles its way round this, by making ```data``` be either bytes or string in a way that matches what ```output``` expects. But it's complicated, fragile code that doesn't keep a clean type model (something that in my experience is the source of a lot of bugs).

I would *strongly* recommend a more type-consistent fix, cleanly separating bytes and strings (conceptually in Python 2, but in terms of actual types in Python 3). That might mean needing to explicitly pass around encodings in places where it's implied at the moment.

But working round the issue in the way your patch does may be a "practicality beats purity" issue - I'll leave that decision to @bitprophet. But someone should at least test it on Python 3 and Windows (and by "test" I mean use it for real, with an actual stdout, not running the tests which stub out the string-based sys.stdout with a bytes-based CarbonCopy instance!!)

Actually, if the test harness is mocking out sys.stdout with a BytesIO object, that's pretty broken. It'll make it pretty much impossible to spot or check issues like this where strings and bytes get muddled.